### ISSUE TITLE
Option to bypass "getOne" step

### ISSUE THREAD

**Issue by coudot** [lsc-project/lsc#163]
For small directories, or when few attributes are synced, we could fetch all values in the "getAll" step and bypass the "getOne" step for each entry in source.

**Comment by coudot** [comment]
This issue was about improving performances on source (requesting all entries at once) but we could also try the same thing with the destination: use the getAll request to load all data in memory instead of requesting each entry. This would be a major improvement for plugins using REST API in destination.

**Comment by elecharny** [comment]
Indeed, but again, the problem is that unless we know the size of the database, we have no way to protect LSC from an OOM. So either we 'poke' the database, fetching a limited number of entries fully (like, 1000), expecting we could get the full database, and if not, then we switch back to a one by one fetch.

Another option would be to fetch only the pivot for all the entries, as we currently do, and if it's below a given number, we fetch all the entries. That makes it two LdapSearches, but we spare many following atomic searches.

**Comment by elecharny** [comment]
Ok, the current implemented strategy is:
* if ( (allFilter == oneFilter)  or (oneFilter == null)) then fetch the entry from source using the listed fetchedAttributes
* then bypass the getEntry() from the source (as we already have them)
* and compare the entry with the getEntry() result from destination.

One thing that needs to be checked is if we want to take into account the dataSets in this very case, but I don't think so. Worst case scenario, use the '*' attribute to get all of them from the source (not a smart idea as we may get back some unused attributes.

Currently, the code that implements this strategy fails on a few tests:
[ERROR] Tests run: 439, Failures: 3, Errors: 0, Skipped: 0

I'm investigating.

Then next improvement, to avoid getting all the read source entries in memory is to use a pagined search.

**Comment by elecharny** [comment]
The errors are due to the fact the `SimpleJndiSrcService.getBean()` function is also visible, and use din the tests.
The side effect of checking that oneFilter is equal to allFilter in this method is that we wrongly assume we already have fetched the entry, which is never the case when calling it directly from the service.
The result is that we get back the wrong values.

**Comment by coudot** [comment]
See https://github.com/lsc-project/lsc/commit/674f7b52d244e5fce7713ac636af47f243e61b98

**Comment by davidcoutadeur** [comment]
I think the best way for this feature is to add a new boolean parameter in service (source or destination):

```
bypassOneEntry
```

when this param is true, LSC should get all fetched attributes during initial request. Need some time to try a poc for this feature.

**Comment by elecharny** [comment]
To rewrite

**Comment by davidcoutadeur** [comment]
It would have been too much work to have the ``bypassOneEntry`` param working for all kinds of directions (source and destination), and services (LDAP, JDBC, plugins,...) . Indeed, for destination it would require to find a way to apply the LDAP filter for matching the destination entries fetched in memory. And for the JDBC and plugins services, we would have rewritten all the logic. So for having the complete feature, it would require a big refactoring.

Thus, it has been decided to propose the feature only for LDAP source services.

- [x] The PR has been done in #532
- [x] the documentation has been written here: https://github.com/lsc-project/documentation/commit/d78dba06d11d0f2b156eea5a2cf8e22e8e33c2e4
- [ ] @elecharny I let you make a final review of PR #532, and I think we can merge safely.
